src/Renderer/NavigatorRenderer.cpp

```diff
@@ -0,0 +1,378 @@
+#include "Formatter/TimeFormatter.hpp"
+#include "ui/canvas/Font.hpp"
+#include "ui/dim/Rect.hpp"
+
+
+// standard
+#include <iomanip>
+#include <iostream>
+#include <string>
+
+
+void
+NavigatorRenderer::DrawFrame(Canvas &canvas, const PixelRect &rc, const NavigatorLook &look) {
+  const auto top_left = rc.GetTopLeft();
+  const int rc_height = rc.GetHeight();
+  const int rc_width = rc.GetWidth();
+  
+  BulkPixelPoint polyline_frame[11];
+  polyline_frame[0].x = top_left.x + 5 / 20.0 * rc_height; //
+  polyline_frame[0].y = top_left.y; //
+  polyline_frame[1].x = top_left.x + rc_width - 4 / 20.0 * rc_height; //
+  polyline_frame[1].y = polyline_frame[0].y; //
+  polyline_frame[2].x = top_left.x + rc_width - 2 / 20.0 * rc_height; //
+  polyline_frame[2].y = top_left.y + 1 / 20.0 *rc_height;; //
+  polyline_frame[3].x = top_left.x + rc_width; //
+  polyline_frame[3].y = top_left.y + 10 / 20.0 * rc_height; //
+  polyline_frame[4].x = polyline_frame[2].x; //
+  polyline_frame[4].y = top_left.y + 19 / 20.0 *rc_height; //
+  polyline_frame[5].x = polyline_frame[1].x; //
+  polyline_frame[5].y = top_left.y + rc_height; //
+  polyline_frame[6].x = polyline_frame[0].x; //
+  polyline_frame[6].y = polyline_frame[5].y; //
+  polyline_frame[7].x = top_left.x + 2 / 20.0 * rc_height ; //
+  polyline_frame[7].y = polyline_frame[4].y; //
+  polyline_frame[8].x = top_left.x; //
+  polyline_frame[8].y = polyline_frame[3].y;//
+  polyline_frame[9].x = polyline_frame[7].x; //
+  polyline_frame[9].y = polyline_frame[2].y; //
+  polyline_frame[10].x = polyline_frame[0].x; //
+  polyline_frame[10].y = polyline_frame[0].y; //
+  
+  canvas.Select(look.background_pen);
+  canvas.Select(look.background_brush);
+  canvas.DrawPolygon( polyline_frame, 11);
+
+  canvas.Select(look.frame_pen);
+  canvas.Select(look.frame_brush);
+  canvas.DrawPolyline( polyline_frame, 11);
+}
+
+
+void 
+NavigatorRenderer::DrawText(Canvas &canvas, const Waypoint &wp_current, 
+                                const PixelRect &rc, const NavigatorLook &look) {
+
+  const auto &basic = CommonInterface::Basic();
+  const auto &calculated = CommonInterface::Calculated();
+  TCHAR current_speed[256] = "---";
+
+  if (basic.ground_speed_available)
+    FormatSpeed(current_speed, Units::ToSysUnit(basic.ground_speed, Unit::METER_PER_SECOND),
+              Unit::KILOMETER_PER_HOUR, true, true);
+  
+  auto current_speed_s = static_cast<std::string>(current_speed);
+    
+  TextRenderer text_renderer1;
+  text_renderer1.SetVCenter(true);
+  text_renderer1.SetCenter(false);
+  text_renderer1.SetControl(true);
+
+  TextRenderer text_renderer2;
+  text_renderer2.SetVCenter(true);
+  text_renderer2.SetCenter(true);
+  text_renderer2.SetControl(true);
+  
+  canvas.Select(*look.font);
+
+  bool has_started = calculated.ordered_task_stats.start.task_started;
+
+  const auto time_elapsed = TimeStamp{FloatDuration{calculated.ordered_task_stats.total.time_elapsed}};
```

Comment:
You shouldn't use the `TimeStamp` type for representing a duration. `TimeStamp` is a time point, not a duration.

Reply:
I didn't have time to check for other representations.
I will do it in the next days.

but if I watch the class definition, it is written:
```
/**
 * This class describes a time point during a flight.  It is used for
 * XCSoar's calculations.  Internally, it is the duration since
 * midnight of the day the flight started.
 */
class TimeStamp {
  FloatDuration value;
```
and then it is very useful to use the function "FormatLocalTimeHHMM" giving directly the good time or elapsed time:

do you have any idea ?